test/token_holder/redeem.js

```diff
@@ -0,0 +1,833 @@
+    },
+    {
+      t: 'bytes4', v: getRedeemCallPrefix(),
+    },
+    {
+      t: 'uint8', v: 0,
+    },
+    {
+      t: 'bytes32', v: '0x0',
+    },
+  );
+
+  const rsv = EthUtils.ecsign(
+    EthUtils.toBuffer(msgHash),
+    EthUtils.toBuffer(ephemeralKey),
+  );
+
+  return { msgHash, rsv };
+}
+
+function getRedeemDataToSign(
+  amount, beneficiary, gasPrice, gasLimit, redeemerNonce) {
+  return web3.eth.abi.encodeFunctionCall(
+    {
+      name: 'redeem',
+      type: 'function',
+      inputs: [
+        {
+          type: 'uint256',
+          name: 'amount',
+        },
+        {
+          type: 'address',
+          name: 'beneficiary',
+        },
+        {
+          type: 'uint256',
+          name: 'gasPrice',
+        },
+        {
+          type: 'uint256',
+          name: 'gasLimit',
+        },
+        {
+          type: 'uint256',
+          name: 'redeemerNonce',
+        },
+      ],
+    },
+    [amount, beneficiary, gasPrice, gasLimit, redeemerNonce],
+  );
+}
+
+async function prepareRedeemPayableRule(
+    accountProvider,
+    amount,
+    beneficiary,
+    facilitator,
+    gasPrice,
+    gasLimit,
+    redeemerNonce,
+    tokenHolder,
+    nonce,
+    ephemeralKey,
+)
+{
+  const mockRule = await MockRule.new();
+
+  const coGatewayRedeemEncodedAbiWithoutHashLock = await getRedeemDataToSign(
+      amount,
+      beneficiary,
+      gasPrice,
+      gasLimit,
+      redeemerNonce
+    );
+
+  const { msgHash, rsv } = getRedeemSignedData(
+    tokenHolder.address,
+    mockRule.address,
+    coGatewayRedeemEncodedAbiWithoutHashLock,
+    nonce,
+    ephemeralKey,
+  );
+
+  return {
+    mockRule,
+    msgHash,
+    rsv,
+  };
+}
+
+async function setupTokenHolder(
+  accountProvider, _ephemeralKeyAddress, _spendingLimit, _deltaExpirationHeight,
+) {
+  token = await UtilityTokenMock.new(1, 1, 'OST', 'Open Simple Token', 1);
+  const tokenRules = await TokenRulesMock.new();
+  const required = 1;
+  const registeredWallet0 = accountProvider.get();
+  const wallets = [registeredWallet0];
+
+  const tokenHolder = await TokenHolder.new(
+    token.address,
+    tokenRules.address,
+    wallets,
+    required,
+  );
+
+  const blockNumber = await web3.eth.getBlockNumber();
+
+  await tokenHolder.submitAuthorizeSession(
+    _ephemeralKeyAddress,
+    _spendingLimit,
+    blockNumber + _deltaExpirationHeight,
+    { from: registeredWallet0 },
+  );
+
+  return {
+    tokenHolder,
+    registeredWallet0,
+  };
+}
+
+contract('TokenHolder::redeem', async (accounts) => {
+
+  describe('Negative Tests', async () => {
+    const accountProvider = new AccountProvider(accounts);
+    const spendingLimit = 50,
+      deltaExpirationHeight = 100,
+      amount = 10,
+      beneficiary = accountProvider.get(),
+      facilitator = accountProvider.get(),
+      gasPrice = 10,
+      gasLimit = 10,
+      redeemerNonce = 1,
+      hashLock = web3.utils.soliditySha3('hl'),
+      payableValue = 100,
+      nonce = 1;
```

Comment:
Where reasonable, I think tests should obtain information the way the software is expected to be used, and not omisciently (i.e., simply knowing the nonce with certainty vs. using the getters).

It would likely be an unreasonable amount of effort to mock getting the redeemerNonce (here, a comment would do), but it should be an easy lift to get the nonce.

Reply:
Just to avoid confusion redeemerNonce  is not stored in TokenHolder. TokenHolder doesn't care anything about redeemerNonce. TH just relays the redeemerNonce argument to CoGateway.redeem. redeemerNonce is stored in CoGateway message object.
That's why the mocking in tests needed.